### PR TITLE
DATA: Add PostImpression Model

### PR DESCRIPTION
**Pull request by tsatay**

### CODE REVIEW

File: Taarafo.Core/Models/PostImpressions/PostImpression.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+        public Guid ProfileId { get; set; }
+        public Profile Profile { get; set; }
+
+        public DateTimeOffset CurrentDate { get; set; }
+        public DateTimeOffset UpdateDate { get; set; }
+
+        public ImpressionType Impression { get; set; }
+
+
```
Comment (hassanhabib) [inline Taarafo.Core/Models/PostImpressions/PostImpression.cs:27]: @tsatay remove these empty lines.

**Comment by hassanhabib** [comment]
@tsatay we still need to set foreign keys between `Profile` and `PostImpression`
And also between `Post` and `PostImpression`

here's an example of how to do that:
In the very first 20 mins I setup a relationship between a `LibraryCard` and a `LibraryAccount` it's the exact same thing when setting up the relationship between a `PostImpression` and a `Post` and the same thing for `Profile` as well.

File: Taarafo.Core/Brokers/Storages/StorageBroker.cs [hunk 2]
```diff
@@ -22,8 +22,13 @@ public StorageBroker(IConfiguration configuration)
             this.Database.Migrate();
         }
 
-        protected override void OnModelCreating(ModelBuilder modelBuilder) =>
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
             AddCommentReferences(modelBuilder);
+            AddPostImpressionsReferences(modelBuilder);
+        }
+
```
Comment (hassanhabib) [inline Taarafo.Core/Brokers/Storages/StorageBroker.cs:30]: Just one line between the two method ya @tsatay

File: Taarafo.Core/Brokers/Storages/IStorageBroker.PostImpressions.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+// ---------------------------------------------------------------
+
+using System.Threading.Tasks;
+using Taarafo.Core.Models.PostImpressions;
+
+namespace Taarafo.Core.Brokers.Storages
+{
+    public partial interface IStorageBroker
+    {
+        ValueTask<PostImpression> InsertPostImpressionAsync(PostImpression postImpression);
```
Comment (hassanhabib) [inline Taarafo.Core/Brokers/Storages/IStorageBroker.PostImpressions.cs:13]: We don't need that. Just the Model & Migrations that's all.

File: Taarafo.Core/Models/PostImpressions/PostImpression.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+using System;
+using Taarafo.Core.Models.Posts;
+using Taarafo.Core.Models.Profiles;
+
+namespace Taarafo.Core.Models.PostImpressions
+{
+    public class PostImpression
+    {
+        public Guid Id{ get; set; }
```
Comment (hassanhabib) [inline Taarafo.Core/Models/PostImpressions/PostImpression.cs:14]: Let's consider making the `PostId` and `ProfileId` the composite primary key for this value.

Which means, you will need an additional step in your reference setup that looks as follows:

```csharp
 modelBuilder.Entity<PostImpression>()
        .HasKey(c => new { c.PostId, c.ProfileId });
```

File: Taarafo.Core/Models/Posts/Post.cs [hunk 5]
```diff
@@ -18,5 +19,7 @@ public class Post
         public DateTimeOffset UpdatedDate { get; set; }
 
         public IEnumerable<Comment> Comments { get; set; }
+        public IEnumerable<PostImpression> PostImpressions { get; set; }
     }
+
```
Comment (hassanhabib) [inline Taarafo.Core/Models/Posts/Post.cs:24]: Don't need this extra line

File: Taarafo.Core/Brokers/Storages/StorageBroker.PostImpressions.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Taarafo.Core.Models.PostImpressions;
+
+namespace Taarafo.Core.Brokers.Storages
+{
+    public partial class StorageBroker
+    {
+        public DbSet<PostImpression> PostImpressions { get; set; }
+
+        public async ValueTask<PostImpression> InsertPostImpressionAsync(PostImpression postImpression)
```
Comment (hassanhabib) [inline Taarafo.Core/Brokers/Storages/StorageBroker.PostImpressions.cs:17]: We don't need this guy

File: Taarafo.Core/Brokers/Storages/IStorageBroker.PostImpressions.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+// ---------------------------------------------------------------
+
+using System.Threading.Tasks;
+using Taarafo.Core.Models.PostImpressions;
+
+namespace Taarafo.Core.Brokers.Storages
+{
+    public partial interface IStorageBroker
+    {
+
```
Comment (hassanhabib) [inline Taarafo.Core/Brokers/Storages/IStorageBroker.PostImpressions.cs:13]: Remove this extra line

File: Taarafo.Core/Brokers/Storages/StorageBroker.PostImpressions.References.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+                .HasForeignKey(postImpression => postImpression.PostId)
+                .OnDelete(DeleteBehavior.NoAction);
+
+            modelBuilders.Entity<PostImpression>()
+                .HasOne(postImpression => postImpression.Profile)
+                .WithOne(profile => profile.PostImpression)
+                .OnDelete(DeleteBehavior.NoAction);
+
+            modelBuilders.Entity<PostImpression>()
+                .HasKey(c => new { c.PostId, c.ProfileId });
```
Comment (hassanhabib) [inline Taarafo.Core/Brokers/Storages/StorageBroker.PostImpressions.References.cs:27]: Move this guy to the top because we have to define that first.

Comment (hassanhabib) [reply Taarafo.Core/Brokers/Storages/StorageBroker.PostImpressions.References.cs:27]: Also change the `c` to `postImpression`

File: Taarafo.Core/Brokers/Storages/StorageBroker.PostImpressions.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Taarafo.Core.Models.PostImpressions;
+
+namespace Taarafo.Core.Brokers.Storages
+{
+    public partial class StorageBroker
+    {
+        public DbSet<PostImpression> PostImpressions { get; set; }
+
```
Comment (hassanhabib) [inline Taarafo.Core/Brokers/Storages/StorageBroker.PostImpressions.cs:16]: Remove empty line

File: Taarafo.Core/Models/PostImpressions/PostImpression.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+using System;
+using Taarafo.Core.Models.Posts;
+using Taarafo.Core.Models.Profiles;
+
+namespace Taarafo.Core.Models.PostImpressions
+{
+    public class PostImpression
+    {
+
```
Comment (hassanhabib) [inline Taarafo.Core/Models/PostImpressions/PostImpression.cs:14]: Remove extra line

File: Taarafo.Core/Models/Profiles/Profile.cs [hunk 11]
```diff
@@ -15,5 +16,7 @@ public class Profile
         public string Email { get; set; }
         public DateTimeOffset CreatedDate { get; set; }
         public DateTimeOffset UpdatedDate { get; set; }
+
+        public PostImpression PostImpression { get; set; }
```
Comment (hassanhabib) [inline Taarafo.Core/Models/Profiles/Profile.cs:20]: This guy needs to be a list

File: Taarafo.Core/Brokers/Storages/StorageBroker.PostImpressions.References.cs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        {
+            modelBuilders.Entity<PostImpression>()
+                .HasOne(postImpression => postImpression.Post)
+                .WithMany(post => post.PostImpressions)
+                .HasForeignKey(postImpression => postImpression.PostId)
+                .OnDelete(DeleteBehavior.NoAction);
+
+            modelBuilders.Entity<PostImpression>()
+                .HasOne(postImpression => postImpression.Profile)
+                .WithOne(profile => profile.PostImpression)
```
Comment (hassanhabib) [inline Taarafo.Core/Brokers/Storages/StorageBroker.PostImpressions.References.cs:23]: This guy should say `.WithMany`

File: DropAndCreate dbo.PostImpressions_1.sql [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+GO
+CREATE TABLE [dbo].[PostImpressions] (
+    [PostId]      UNIQUEIDENTIFIER   NOT NULL,
+    [ProfileId]   UNIQUEIDENTIFIER   NOT NULL,
+    [CurrentDate] DATETIMEOFFSET (7) NOT NULL,
+    [UpdateDate]  DATETIMEOFFSET (7) NOT NULL,
+    [Impression]  INT                NOT NULL
+);
+
```
Comment (hassanhabib) [inline DropAndCreate dbo.PostImpressions_1.sql:22]: So, why do we need that? :)

Comment (tsatay) [reply DropAndCreate dbo.PostImpressions_1.sql:22]: I had to drop and create the table from the database to regenerate the migration for Post Impression.

**Comment by hassanhabib** [comment]
@tsatay as we discussed, just start over.

**Pull request closed without merging**